### PR TITLE
WIP: Nodeset compiler for v0.3

### PR DESCRIPTION
**Pull request by jpfr**
This PR merges #725 and #729.
Furthermore, I started to refactor the open62541 code generation backend into a separate module.
So the compiler stages are represented in the code architecture.

### CODE REVIEW

**Comment by markusgraube** [comment]
@jpfr The restructuring looks interesting. However, I don't understand where the bootstrapping code has gone. Is this not necessary anymore?

**Comment by jpfr** [comment]
This PR contains only the node set compiler part. The part concerning the ns0 bootstrapping in the C-code is in the `nodeset_compiler_02_server` branch.

The goal of this PR is
- Refactor with more object-orientation (classes, inheritance) and clear separation of "compiler" stages in the code
- Solve all open issues for the nodeset compiler flagged for 0.2
- Speed up the nodeset compiler. It now takes less than 1s to generate the **full ns0** and much less for the minimal nodeset. Before, we were at ~15 seconds thanks to the ordering algrithm from @Stasik0!
- A unified solution for ns0 bootstrapping and normal information model generation

The code generation only uses the public API. However, for creating ns0, we need to circumvent some consistency checks that would fail (e.g. root has no parent node).
That is done here: https://github.com/open62541/open62541/blob/nodeset_compiler_02_server/src/server/ua_server.c#L280
` ua_namespace0` is the generated function that adds all ns0 nodes and references.

Some stuff needs to be brought back in here before merging.
- The changes you did for the high-level API
- Some of @ichrispa's original functionality is not fully working in this branch yet. E.g. creating extensionobjects with structures as content

Once this is done, we can use the generated ns0 in the standard build.

**Comment by jpfr** [comment]
By the way... this is a **refactoring, not a rewrite**.
Writing the thing from scratch is like 10x the effort.
We wouldn't have any nodeset compiler without @ichrispa 's great contribution.
Sadly he indicated to have no time for maintenance right now.

After a period of large growth refactoring is necessary almost always.
Otherwise, technical debt accumulates. (Shortcuts taken during development, overhead for data structures that were once important but are used only rarely now, similar pieces of functionality that merit a general solution, etc.)

Such a refactoring step has happened in most parts of open62541. Sometimes several times before converging to some "local optimum".

**Comment by markusgraube** [comment]
@jpfr Can I somehow help with this PR?

**Comment by jpfr** [comment]
In general, yes of course!
But the PR will not hit 0.2. It's would take too long to restore all current features...

An important part that is missing from the refactoring are the values of variables in a nodeset xml.
If we only had a generic way of adding a (structured) value to an addvariablenode call.
Maybe it's enough to go by the member names of the structure to reconstruct in C...

**Comment by Pro** [comment]
@jpfr should I have a look at setting the variable values you mentioned in the previous commit?
I would first try to use the member names to see if we can manage it with that.

Is there anything else we need for this PR?

**Comment by jpfr** [comment]
>  I would first try to use the member names to see if we can manage it with that.

Sorry, I could not parse that. Can you rephrase?

There is one feature lacking compared to the namespace compiler on master: custom structures.
@ichrispa's approach was to manually encode the struct members into a bytestring and use an extensionobject to hold the (encoded) value.
This works just fine.

Another approach is to generate the struct definition plus the `UA_DataType` and to generate a "real" value. But we might need some more infrastructure. For example adding `UA_DataType` to a server outside of the config struct.

I'm okay with the encoded-extensionobject approach in order to get a non-experimental namespace compiler. If you want a full ns0 bootstrap from XML, we can talk next week about the internal API that needs to be exposed to the namespace compiler.

**Comment by Pro** [comment]
Regarding the values I will try to implement it using the extension object approach, or see what the best solution is.

@jpfr The current version in this PR thows an exception if the Node has no parent:
https://github.com/open62541/open62541/blob/6238408730f6e106243227cf56e834f332f1e719/tools/nodeset_compiler/backend_open62541_nodes.py#L36

In the NodeSet2.xml there are some nodes which do not have any parent. The most prominent one ist the Root node, but there are also others. See the full list attached.

How should we handle those? Just ignore them?

```
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2341) (Interpolative) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12207) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=886) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=885) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=889) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=888) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=861) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=860) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=863) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=867) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=866) (Default XML) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=3069) (LocalTime) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=3068) (NodeVersion) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=3063) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=3062) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=3067) (Icon) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=7616) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=587) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=84) (Root) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=83) (ExposesItsArray) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=80) (Optional) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=588) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2361) (DurationBad) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2360) (DurationGood) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=297) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12901) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12900) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=12908) (MaxByteStringLength) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=854) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=855) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=3070) (AllowNulls) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=3071) (EnumValues) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=3072) (InputArguments) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=3073) (OutputArguments) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=323) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=321) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=539) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=538) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=339) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=333) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=332) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=542) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=541) (Default XML) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=11214) (Annotations) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=11215) (HistoricalEventFilter) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=869) (Default XML) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=11432) (EnumStrings) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=949) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=940) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=309) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=305) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=306) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=300) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=301) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=79) (MandatoryShared) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=78) (Mandatory) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2347) (Maximum) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=8251) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11202) (HA Configuration) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=950) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=313) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=310) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=317) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=314) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=318) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12174) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12173) (Default XML) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=12170) (ViewVersion) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11308) (DurationInStateNonZero) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11304) (Total2) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11305) (MinimumActualTime2)has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11306) (MaximumActualTime2)has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11307) (DurationInStateZero) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12766) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12765) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=377) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=378) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=591) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=597) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12195) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12758) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12757) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11950) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11957) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12680) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11958) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=584) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=585) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=11433) (ValueAsText) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=8913) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=8917) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=340) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=345) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=346) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11508) (OptionalPlaceholder) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11506) (EndBound) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11507) (DeltaBounds) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11505) (StartBound) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12182) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12181) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=12745) (OptionSetValues) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=590) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=593) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=594) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=596) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=599) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11510) (MandatoryPlaceholder) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=11513) (EngineeringUnits)has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=11512) (MaxArrayLength) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11949) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=726) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=727) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=720) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=721) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=433) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=434) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11429) (VariancePopulation)has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11428) (VarianceSample) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11427) (StandardDeviationPopulation) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11426) (StandardDeviationSample) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2364) (WorstQuality) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2363) (PercentBad) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2362) (PercentGood) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=298) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=921) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=922) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11292) (WorstQuality2) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=600) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=603) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=602) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=939) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11288) (Range2) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11285) (TimeAverage2) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11287) (Maximum2) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=11286) (Minimum2) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=858) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=320) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=326) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=327) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=324) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12089) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12082) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12081) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=857) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=661) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=660) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12893) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12892) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=384) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=386) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=387) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=383) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2349) (MaximumActualTime) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2348) (MinimumActualTime) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2343) (TimeAverage) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2344) (Total) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2346) (Minimum) has no parent
INFO:backend_open62541_nodes:Node VariableNode(ns=0;i=11498) (MaxStringLength) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12090) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=12676) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=864) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2358) (End) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2359) (Delta) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2357) (Start) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2355) (NumberOfTransitions) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2352) (Count) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2350) (Range) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2351) (AnnotationCount) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=2342) (Average) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=892) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=893) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=895) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=896) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=898) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=899) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=878) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=879) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=872) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=873) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=870) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=876) (Default Binary) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=875) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=380) (Default XML) has no parent
INFO:backend_open62541_nodes:Node ObjectNode(ns=0;i=381) (Default Binary) has no parent
```

**Comment by Pro** [comment]
@jpfr, @ichrispa can you please review.

I added support for setting the variable values mainly based on the code from @ichrispa and adapted it to the current new Object-oriented structure.

Note that this PR only includes the changes to the Nodeset Compiler. All other changes regarding the server bootstrapping are in #1179 which is based on this.

I had to extend the `Opc.Ua.NodeSet2.Minimal.xml` to include some missing referenced nodes. Maybe it is not the really minimal set we could have, but for now it is working like this.

My suggestion is to merge this PR when #1179 is ready. And then immediately followed by #1179.

**Comment by jpfr** [comment]
Which should go first? This PR or #1168?

**Comment by Pro** [comment]
If #1168 is ready to be merged than that can be first and then it would be nice if you (@jpfr) can rebase this PR onto the new master (it is currently already based on master HEAD, thus not that much effort) since you know your #1168 PR best.

#739 is not yet fully ready to be merged since I want to carry out some more tests with more NodeSet2.xml files and improve the unit tests to avoid further code breaks.

**Comment by Stasik0** [comment]
97??? W0000t? Coverity bug or for real?

**Comment by Pro** [comment]
For real 👍 
It looks like coveralls is also including the generated namespace code, which are a few ten thousand lines covered by 100%, including these lines in the overall percentage causes a drastical increase. Anyways it's still a nice number 😃

**Comment by Pro** [comment]
All green, ALL GREEN :white_check_mark: :cool: :champagne: 

All the continuous integration tasks run through without any error.
The code now also includes a unit test which compiles the full NS0, then the DI model followed by the ADI model.

And a very nice side-effect: Coveralls now reports a code coverage of 96% (probably because it is also taking into account the generated nodeset code, which is a major percentage of the code)

See here for the example on how to do that:
https://github.com/open62541/open62541/blob/5cebb35d396513275ea5d6949ab6aba88c025010/tests/CMakeLists.txt#L196

What still needs to be done is to check the generated NS0 using CTT (@jpfr)

From my point of view this PR is ready. Before we merge it we should have the (not yet) planned telco regarding the nodeset compiler. I'll send the invitation to the relevant people as discussed last time. If someone else is interested in joining the telko please leave a note here.

**Comment by Pro** [comment]
This PR is now based on #1202.

#1202 contains changes which were previously part of this PR and can be merged before the new nodeset compiler.

This PR now only contains necessary changes for the new nodeset compiler which makes review easier.

After merging #1202 we can set the base of this PR back to master.

**Comment by Pro** [comment]
@jpfr can you merge/rebase this on master to fix the merge conflicts. You have a better knowledge what was changed with the nodeset plugin PR.

**Comment by Pro** [comment]
Closing in favor of #1238
